civievent-widget.php

```diff
@@ -207,134 +207,45 @@ public function widget( $args, $instance ) {
 
 		if ( version_compare( $this->_civiversion, '4.3.alpha1' ) < 0 ) { return; }
 
-		$fields = $this->getFields();
+		$customDisplay = $this->validCustomDisplayFields( $instance );
+		$customDisplayFields = array_intersect_key( self::getCustomDisplayTitles(), $customDisplay );
+
+		$defaultParams = array(
+			'start_date' => array( '>=' => date( 'Y-m-d' ) ),
+			'is_public' => 1,
+			'options' => array(
+				'sort' => 'start_date ASC',
+				'limit' => CRM_Utils_Array::value( 'limit', $instance, 5 ),
+			),
+		);
+		$eventTypeIdParams = $this->buildEventTypeIdParams( $instance );
+		$customFilterParams = $this->validCustomFilterFields( $instance );
+		$returnParams = $this->buildReturnParams( $customDisplay, $customDisplayFields );
+		$filterParams = array_merge_recursive( $defaultParams, $eventTypeIdParams, $customFilterParams, $returnParams );
```

Comment:
Using `array_merge_recursive()` is problematic here because it's not smart enough to know when to merge and when to replace. For example:

```


$defaultParams = array(
  'start_date' => array('>=' => '2019-03-19'), // hard-coding today's date for example purposes.
  // ... omitting other params for simplicity ...
);
$customFilterParams = array(
  'start_date' => array('>=' => '2020-01-01'),
);
var_dump(array_merge_recursive($defaultParams, $customFilterParams));
/* Ouput:
 array(1) {
  'start_date' =>
  array(1) {
    '>=' =>
    array(2) {
      [0] =>
      string(10) "2019-03-19"
      [1] =>
      string(10) "2020-01-01"
    }
  }
}
 * This will cause civicrm api to ignore the 'start_date' param entirely.
*/
$customFilterParams = array(
  'start_date' => array('=' => '2020-01-01'),
);
var_dump(array_merge_recursive($defaultParams, $customFilterParams));
/* Ouput:
array(1) {
  'start_date' =>
  array(2) {
    '>=' =>
    string(10) "2019-03-19"
    '=' =>
    string(10) "2020-01-01"
  }
}
 * This will cause civicrm api to return events matching EITHER of the 
 * defined 'start_date' criteria.
*/
```

I did have some hope for `array_replace_recursive()`, but it's also not smart enough, and will also result in weird combinations.

I think the only way to overcome this is with an additional method to merge these parameters explicitly to get the intended parameters.

Reply:
OK, actually, I think there's hope for simply `array_merge()` here. It  resolves the above-mentioned issues, but not sure if it causes other issues.